storage/remote/itern.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Inspired / copied / modified from https://gitlab.com/cznic/strutil/blob/master/strutil.go,
+// which is MIT licensed, so:
+//
+// Copyright (c) 2014 The strutil Authors. All rights reserved.
+
+package remote
+
+import "sync"
+
+var iterner = newPool()
+
+type pool struct {
+	mtx  sync.RWMutex
+	pool map[string]string
+}
+
+func newPool() *pool {
+	return &pool{
+		pool: map[string]string{},
+	}
+}
+
+func (p *pool) intern(s string) string {
+	if s == "" {
+		return ""
+	}
+
+	p.mtx.RLock()
+	interned, ok := p.pool[s]
+	p.mtx.RUnlock()
+	if ok {
+		return interned
+	}
+
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+	if interned, ok := p.pool[s]; ok {
```

Comment:
This is the standard read-write mutex pattern; the first check is under the read lock and therefore concurrent, but you still need a second check it under write lock just in case someone got in there first and added the string between you dropping the read lock and taking the write lock.